### DISCUSSION TITLE
easywechat access_token，怎么切换成redis存储

### DISCUSSION THREAD

**Discussion by TheSaltwaterRoom** [w7corp/easywechat#2738] [Q&A]
1.如果access_token过期后，重刷机制有人搞过吗，怎么做
2.6.x版本，有的路由上面，需要拼接access_token,我用 $client->withAccessToken 这个方法不行，

**Comment by wjfz** [comment]
0. 切换 Redis 存储，大概这么弄：(new Application($config))->setCache(new Psr16Cache(new RedisAdapter(app('redis')->connection()->client())));
1. AccessToken->getToken() 会自动重刷，并且有个 AccessToken->refresh()
2. $client->withAccessToken(AccessTokenInterface) 传个AccessToken对象进去就行，不过一般不需要这么做。

**Reply by TheSaltwaterRoom** [reply]
比如下面的例子
$response = $api->postJson('/cgi-bin/qrcode/create?access_token=xxxx', [

改为

$response = $api->withAccessToken($accessToken)->postJson('/cgi-bin/qrcode/create', [

然后就报错了

{"errcode":40001,"errmsg":"invalid credential, access_token is invalid or not latest, could get access_token by getStableAccessToken, more details at https://mmbizurl.cn/s/JtxxFh33r rid: 64f1b3af-66d17331-26ee0c5d"}

**Reply by TheSaltwaterRoom** [reply]
还有就是自动重刷的话，为什么还会报 40001 这个错误

**Reply by wjfz** [reply]
如果上面的方法可以执行，但是下面的报错了，就打印下$accessToken->getToken()，看看$accessToken取的对不对。

**Reply by TheSaltwaterRoom** [reply]
这个自动重刷机制，他是设置了一个缓存，缓存取不到，就重新去查，像这种情况，应该不会出现40004这种错误吧

**Comment by gaoshulin** [comment]
负载均衡服务器，改为 redis 存储，亲测有效，laravel 项目实现如下：

// 实例化 Application
$this->app = new \EasyWeChat\OfficialAccount\Application(config('easywechat.official_account'));

 // 使用 redis 缓存
 $redis = Redis::connection('cache')->client();
$this->app->setCache(new Psr16Cache(new RedisAdapter($redis, namespace: 'easywechat', defaultLifetime: 3600)));
